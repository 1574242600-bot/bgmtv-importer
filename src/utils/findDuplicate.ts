/**
 * 查找并返回重复的元素及其索引
 * 
 * @name findDuplicate
 * @param array T[] 
 * @param valueOf (value: T) => string 一个函数,参数是数组中的元素,返回字符串,用作判断元素是否重复的依据
 * @return Record<string, number[]> string 是 valueOf() 返回值, number[] 索引. 未找到重复的元素时,返回null
 */
export default function findDuplicate<T>(
    array: T[],
    valueOf: (value: T) => string,
): Record<string, number[]> {
    const firstIndex: Record<string, number> = {};

    const result = array.reduce<Record<string, number[]>>((duplicate, value, index) => {
        const key = valueOf(value);
        if (firstIndex[key] === undefined) {
            firstIndex[key] = index;
            return duplicate;
        }

        if (duplicate[key] === undefined) {
            duplicate[key] = [];
            duplicate[key].push(firstIndex[key]);
            duplicate[key].push(index);
            return duplicate;
        }

        //if (duplicate[key] instanceof Array)
        duplicate[key].push(index);

        return duplicate;
    }, {});

    return Object.keys(result).length === 0 ? null : result;
}