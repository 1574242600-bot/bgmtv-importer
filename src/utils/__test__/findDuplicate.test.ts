import findDuplicate from '../findDuplicate';

test('找到重复项, 返回其索引信息',() => {
    const received = findDuplicate([1,2,2,3], (v) => v + '');
    expect(received).toMatchSnapshot();
});

test('未找到重复项, 返回null',() => {
    const received = findDuplicate([1,2,3], (v) => v + '');
    expect(received).toBe(null);
});
