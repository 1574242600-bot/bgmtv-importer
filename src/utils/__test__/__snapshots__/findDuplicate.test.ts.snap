// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`找到重复项, 返回其索引信息 1`] = `
Object {
  "2": Array [
    1,
    2,
  ],
}
`;
